sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs

```diff
@@ -0,0 +1,100 @@
+    public sealed class CollectionsShouldImplementGenericInterface : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3909";
+        private const string MessageFormat = "Refactor this collection to implement '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly Dictionary<KnownType, KnownType> nongenericToGenericMapping =
+            new Dictionary<KnownType, KnownType>
+            {
+                [KnownType.System_Collections_ICollection] = KnownType.System_Collections_Generic_ICollection_T,
+                [KnownType.System_Collections_IList] = KnownType.System_Collections_Generic_IList_T,
+                [KnownType.System_Collections_IEnumerable] = KnownType.System_Collections_Generic_IEnumerable_T,
+                [KnownType.System_Collections_CollectionBase] = KnownType.System_Collections_ObjectModel_Collection_T,
+            };
+
+        private static readonly ISet<KnownType> nongenericTypes = nongenericToGenericMapping.Keys.ToHashSet();
+        private static readonly ISet<KnownType> genericTypes = nongenericToGenericMapping.Values.ToHashSet();
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+            c =>
+            {
+                var classDeclaration = c.Node as ClassDeclarationSyntax;
+
+                var issues = new List<Diagnostic>();
+
+                var implementedTypes = classDeclaration?.BaseList?.Types;
+                if (implementedTypes == null)
+                {
+                    return;
+                }
+
+                foreach (var typeSyntax in implementedTypes)
+                {
+                    var typeSymbol = c.SemanticModel.GetSymbolInfo(typeSyntax.Type).Symbol?.GetSymbolType();
+
+                    var originalDefinition = typeSymbol?.OriginalDefinition;
+                    if (originalDefinition.IsAny(genericTypes))
+                    {
+                        return;
+                    }
+
+                    var nongenericType = ToKnownType(typeSymbol, nongenericTypes);
+                    if (nongenericType != null)
+                    {
+                        issues.Add(Diagnostic.Create(rule,
+                                    classDeclaration.Identifier.GetLocation(),
+                                    nongenericToGenericMapping[nongenericType].TypeName));
+                    }
+                }
+
+                issues.ForEach(i => c.ReportDiagnostic(i));
+            },
+            SyntaxKind.ClassDeclaration);
+        }
+
+        private static KnownType ToKnownType(ITypeSymbol typeSymbol, IEnumerable<KnownType> knownTypes)
```

Comment:
You could remove the `knownTypes` argument and perhaps rename the method to something like `SuggestGenericCollectionType` or better and refactor it to directly return a generic type counterpart or null... WDYT?